### PR TITLE
fix: Fix switching viz type to and from Filter box

### PR DESCRIPTION
**Pull request by kgabryje**
### SUMMARY
Switching viz type between a chart type and filter was broken due to a different type of data payload (array for charts and object for filter) and additional rerender caused by dispatching `RESET_FIELDS` after render caused by switching viz type. I moved the logic of resetting control fields to `SET_FIELD_VALUE` reducer, thus removing 1 rerender.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Before: see https://github.com/apache/superset/issues/12624

After:
https://user-images.githubusercontent.com/15073128/107769388-32095280-6d38-11eb-9b1b-f083dba1b1d9.mov

### ADDITIONAL INFORMATION

- [x] Has associated issue: fixes https://github.com/apache/superset/issues/12624
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

CC: @junlincc @villebro @nikolagigic @rusackas

### CODE REVIEW

**Comment by junlincc** [comment]
approving as product sign-off. Thanks for the fix! 🙂 @villebro ready for ya. 

https://user-images.githubusercontent.com/67837651/107899533-637a4d80-6ef3-11eb-8ccf-fb547b5d1922.mov

File: superset-frontend/src/explore/reducers/exploreReducer.js [hunk 1]
```diff
@@ -121,17 +121,35 @@ export default function exploreReducer(state = {}, action) {
...
+                validationErrors: errors,
+              },
+            }
+          : {
+              ...state.controls,
+              [action.controlName]: {
+                ...control,
+                validationErrors: errors,
+              },
+            };
```
Comment (ktmud) [inline superset-frontend/src/explore/reducers/exploreReducer.js:146]: These lines are a little difficult to grok, creating a new variable might make it easier to read:

```js
      const currentControlsState =
        action.controlName === 'viz_type' &&
        action.value !== state.controls.viz_type.value
          ? // rebuild the full control state if switching viz type
            getControlsState(
              state,
              getFormDataFromControls({
                ...state.controls,
                viz_type: control,
              }),
            )
          : state.controls;

      return {
        ...state,
        form_data: new_form_data,
        triggerRender: control.renderTrigger && !hasErrors,
        controls: {
          ...currentControlsState,
          [action.controlName]: {
            ...control,
            validationErrors: errors,
          },
        },
      };
```

**Review by ktmud — APPROVED** [review]
Happen to run into this bug and did some testing. This works perfectly!

LGTM with one minor suggestion on code style.

**Review by villebro — APPROVED** [review]
LGTM, this is great! 🚀

**Review by junlincc — APPROVED** [review]

File: superset-frontend/src/explore/reducers/exploreReducer.js (hunk 1 above)
Comment (kgabryje) [reply superset-frontend/src/explore/reducers/exploreReducer.js]: Good point! Done

**Pull request merged**